File: example/app2/ios/Runner/NativeLogDemo.swift
import Foundation
import os.log

class NativeLogDemo {
  private static let pluginLog = OSLog(subsystem: "com.example.myplugin", category: "general")
  private static let networkLog = OSLog(subsystem: "com.example.network", category: "http")
  private static var timer: Timer?
  private static var counter = 0

  /// Emit a burst of sample native logs using NSLog and os_log.
  /// These simulate what iOS Flutter plugins produce via unified logging.
  static func emitSampleLogs() {
    // NSLog — the classic logging mechanism (visible in idevicesyslog and Console.app)
    NSLog("Native log demo triggered from Flutter")
    NSLog("Plugin v2.1.0 loaded — using NSLog for backward compatibility")

    // os_log — modern unified logging with subsystem/category
    os_log("Plugin initializing native components", log: pluginLog, type: .info)
    os_log("Configuration loaded from bundle", log: pluginLog, type: .debug)
    os_log("Sensor framework connected", log: pluginLog, type: .info)
    os_log("Calibration data stale — recalibrating", log: pluginLog, type: .default)

    // Network subsystem logs
    os_log("GET https://api.example.com/data", log: networkLog, type: .debug)
    os_log("Response: 200 OK (45ms)", log: networkLog, type: .debug)
    os_log("TLS certificate expires in 7 days", log: networkLog, type: .default)

    // Error/fault levels
    os_log("Connection timeout after 30s", log: networkLog, type: .error)
    os_log("Critical: data corruption detected (simulated)", log: pluginLog, type: .fault)

    // Multi-line NSLog (simulated stack trace)
    NSLog("Error in native plugin: NullPointerException")
    NSLog("  at MyPlugin.fetchData() (MyPlugin.swift:42)")
    NSLog("  at PluginHandler.handleCall() (PluginHandler.swift:87)")
  }

  /// Start emitting periodic native logs.
  static func startPeriodicLogs(intervalMs: Int) {
    stopPeriodicLogs()
    counter = 0
    let interval = TimeInterval(intervalMs) / 1000.0

    // Timer must be created on the main thread for RunLoop scheduling
    DispatchQueue.main.async {
      timer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { _ in
        counter += 1
        switch counter % 4 {
        case 0:
          os_log("Heartbeat #%d: connection alive", log: pluginLog, type: .info, counter)
        case 1:
          os_log("Request #%d: GET /api/status -> 200", log: networkLog, type: .debug, counter)
        case 2:
          NSLog("Event #%d: sensor data received", counter)
        default:
          os_log("Tick #%d: background task running", log: pluginLog, type: .debug, counter)
        }
      }
    }
  }

  static func stopPeriodicLogs() {
    timer?.invalidate()
    timer = nil
  }
}

File: example/app2/android/app/src/main/kotlin/com/example/flutter_deamon_sample/NativeLogDemo.kt
package com.example.flutter_deamon_sample

import android.util.Log
import java.util.Timer
import java.util.TimerTask

object NativeLogDemo {
    private const val TAG_DEMO = "NativeDemo"
    private const val TAG_PLUGIN = "MyPlugin"
    private const val TAG_GO = "GoLog"
    private const val TAG_NETWORK = "OkHttp"

    private var periodicTimer: Timer? = null
    private var logCounter = 0

    /**
     * Emit a burst of sample native logs across different tags and priorities.
     * These simulate what a real Flutter app with native plugins would produce.
     */
    fun emitSampleLogs() {
        // Simulate a native plugin initialization sequence
        Log.i(TAG_DEMO, "Native log demo triggered from Flutter")
        Log.d(TAG_PLUGIN, "Plugin initializing native components")
        Log.i(TAG_PLUGIN, "Plugin v2.1.0 loaded successfully")

        // Simulate Go/gomobile output (common in cross-platform Flutter apps)
        Log.d(TAG_GO, "Go runtime initialized, GOMAXPROCS=4")
        Log.i(TAG_GO, "gRPC client connected to backend:8443")
        Log.w(TAG_GO, "TLS certificate expires in 7 days")

        // Simulate network library output
        Log.d(TAG_NETWORK, "--> GET https://api.example.com/data")
        Log.d(TAG_NETWORK, "--> END GET")
        Log.d(TAG_NETWORK, "<-- 200 OK (45ms)")
        Log.d(TAG_NETWORK, "<-- END HTTP (1234-byte body)")

        // Simulate various priority levels
        Log.v(TAG_DEMO, "Verbose: detailed trace information")
        Log.d(TAG_DEMO, "Debug: diagnostic information")
        Log.i(TAG_DEMO, "Info: general information")
        Log.w(TAG_DEMO, "Warning: potential issue detected")
        Log.e(TAG_DEMO, "Error: something went wrong (simulated)")

        // Simulate a multi-line native log (stack trace style)
        Log.e(TAG_PLUGIN, "NullPointerException: Attempt to invoke virtual method on null reference")
        Log.e(TAG_PLUGIN, "  at com.example.plugin.DataManager.fetchData(DataManager.kt:42)")
        Log.e(TAG_PLUGIN, "  at com.example.plugin.PluginHandler.handleCall(PluginHandler.kt:87)")
    }

    /**
     * Start emitting periodic native logs (simulates ongoing plugin activity).
     */
    fun startPeriodicLogs(intervalMs: Long) {
        stopPeriodicLogs()
        logCounter = 0
        periodicTimer = Timer().apply {
            scheduleAtFixedRate(object : TimerTask() {
                override fun run() {
                    logCounter++
                    when (logCounter % 4) {
                        0 -> Log.i(TAG_GO, "Heartbeat #$logCounter: connection alive")
                        1 -> Log.d(TAG_NETWORK, "Request #$logCounter: GET /api/status -> 200 (${(10..200).random()}ms)")
                        2 -> Log.i(TAG_PLUGIN, "Event #$logCounter: sensor data received")
                        3 -> Log.d(TAG_DEMO, "Tick #$logCounter: background task running")
                    }
                }
            }, 0L, intervalMs)
        }
    }

    fun stopPeriodicLogs() {
        periodicTimer?.cancel()
        periodicTimer = null
    }
}
